TicketSwap/solar
# Introduce tests and fix memory leaks of useTransition

# Description

@Roohn and I look into the memory leaks we had in our project tests and found most of them were originated in the `useTransition` hook of the Dialog.

If you have access you can see the warnings categorized [here](https://github.com/TicketSwap/sirius/issues/5839). The improvements this PR introduces reduces the memory leaks from 20 to 5 in the project.

**Important changes made:**
- The transitioning state is removed since we can use the "state" state to determine that;
- `rAF` inside a `rAF` is not necessary if we `useLayoutEffect`. Contrary to `useEffect`, which fires after the browser has painted, `useLayoutEffect` fires before the painting phase of the browser. One `rAF` call is still necessary to schedule the layout update for the current frame. This issue has however the issue of not being available on the server-side and producing an ugly warning. A hack from [here](https://github.com/reduxjs/react-redux/blob/d16262582b2eeb62c05313fca3eb59dc0b395955/src/components/connectAdvanced.js#L40) is applied to silence the warning, but we can keep the previous approach of nested `rAF` if you prefer (let me know in the comments);

https://reactjs.org/docs/hooks-reference.html#uselayouteffect
https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame
https://blog.jakuba.net/request-animation-frame-and-use-effect-vs-use-layout-effect/

- `useCallback` does not have any cleanup functions like `useEffect`, contrary to what we had in the code. Instead, we now clear the timers and scheduled `rAF` when the component unmounts (needs an empty dependency array to fire only once and on unmount).

## Changes


## How to test

**To test the useTransition**
- Checkout this branch
- `$ yarn dev`
- Check the storybook for `ContentTransition`

**To check in your project**
- `$ yarn prepublishOnly`
- Copy the package inside `dist`
- Place it inside the project using solar (`node_modules`)
- Check the transitions

**To check the memory leaks are reduced:**
- Same as above
- Paste this [code](https://gist.github.com/diogoredin/35910541eba889b3da13604307ef187e) in your setup tests file
- Check the debug.log and see the memory leaks are less than stated [here](https://github.com/TicketSwap/sirius/issues/5839). 

## To be notified

@TicketSwap/frontend 

## Links

Fixes #1225
